cmd/kubeadm/app/util/users/users_linux.go

```diff
@@ -0,0 +1,602 @@
+			if i64 == e.id {
+				goto continueLoop
+			}
+		}
+		ids = append(ids, i64)
+		if len(ids) == total {
+			return ids, nil
+		}
+	continueLoop:
+	}
+	return nil, errors.Errorf("could not allocate %d ids based on existing entries in the range: %d - %d",
+		total, min, max)
+}
+
+// addEntries takes /etc/passwd or /etc/group file content and appends entries to it based
+// on a createEntry function. Returns the updated contents of the file.
+func addEntries(file string, entries []*entry, createEntry func(*entry) string) string {
+	out := file
+	newLines := make([]string, 0, len(entries))
+	for _, e := range entries {
+		newLines = append(newLines, createEntry(e))
+	}
+	newLinesStr := ""
+	if len(newLines) > 0 {
+		if !strings.HasSuffix(out, "\n") { // Append a new line if its missing.
+			newLinesStr = "\n"
+		}
+		newLinesStr += strings.Join(newLines, "\n") + "\n"
+	}
+	return out + newLinesStr
+}
+
+// removeEntries takes /etc/passwd or /etc/group file content and deletes entries from them
+// by name matching. Returns the updated contents of the file and the number of entries removed.
+func removeEntries(file string, entries []*entry) (string, int) {
+	lines := strings.Split(file, "\n")
+	total := len(lines) - len(entries)
+	if total < 0 {
+		total = 0
+	}
+	newLines := make([]string, 0, total)
+	removed := 0
+	for _, line := range lines {
+		for _, entry := range entries {
+			if strings.HasPrefix(line, entry.name+":") {
+				removed++
+				goto continueLoop
+			}
+		}
+		newLines = append(newLines, line)
+	continueLoop:
+	}
+	return strings.Join(newLines, "\n"), removed
+}
+
+// assignUserAndGroupIDs takes the list of existing groups, the users and groups to be created,
+// and assigns UIDs and GIDs to the users and groups to be created based on a list of provided UIDs and GIDs.
+// Returns an error if not enough UIDs or GIDs are passed. It does not perform any other validation.
+func assignUserAndGroupIDs(groups, usersToCreate, groupsToCreate []*entry, uids, gids []int64) error {
+	if len(gids) < len(groupsToCreate) {
+		return errors.Errorf("not enough GIDs to assign to groups: have %d, want %d", len(gids), len(groupsToCreate))
+	}
+	if len(uids) < len(usersToCreate) {
+		return errors.Errorf("not enough UIDs to assign to users: have %d, want %d", len(uids), len(usersToCreate))
+	}
+	for i := range groupsToCreate {
+		groupsToCreate[i].id = gids[i]
+	}
+	// Concat the list of old and new groups to find a matching GID
+	groupsConcat := append([]*entry{}, groups...)
+	groupsConcat = append(groupsConcat, groupsToCreate...)
+	for i := range usersToCreate {
+		usersToCreate[i].id = uids[i]
+		for _, g := range groupsConcat {
+			if usersToCreate[i].name == g.name {
+				usersToCreate[i].gid = g.id
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// createGroup is a helper function to produce a group from entry
+func createGroup(e *entry) string {
+	return fmt.Sprintf("%s:x:%d:%s", e.name, e.id, strings.Join(e.userNames, ","))
```

Comment:
since we are using supplementalGroups: we don't need to add the users to the group. Not adding the list of users to the group makes the validation and creation a bit simpler.

Reply:
true, the complexity and number of LoC will be reduced by this.

i can see this not being needed for primary groups (`adduser` does not do that for primary groups, without an explicit `usermod`). but can we also avoid it for `kubeadm-sa-key-readers`, which is supplemental group by definition?

i guess, yes.

Reply:
I guess leaving it is fine, but since the users are already added to the group we wouldn't need to add the supplementalGroup.